sql/parser/eval.go

```diff
@@ -581,6 +605,380 @@ func init() {
 	cmpOps[cmpArgs{In, tupleType, tupleType}] = evalTupleIN
 }
 
+// TypeCheckExpr type-cheks an SQL expression. Expression checking is a
+// mostly straightforward walk over the parse tree.
+// TODO(tamird): tuple handling is currently crappy.
+func TypeCheckExpr(expr Expr) (reflect.Type, error) {
+	switch t := expr.(type) {
+	case Row:
+		// Row and Tuple are synonymous: convert Row to Tuple to simplify logic
+		// below.
+		expr = Tuple(t)
+	}
+
+	switch t := expr.(type) {
+	case *AndExpr:
+		if err := typeCheckBooleanExprs(t.Left, t.Right); err != nil {
+			return nil, err
+		}
+
+		return boolType, nil
+
+	case *OrExpr:
+		if err := typeCheckBooleanExprs(t.Left, t.Right); err != nil {
+			return nil, err
+		}
+
+		return boolType, nil
+
+	case *NotExpr:
+		if err := typeCheckBooleanExprs(t.Expr); err != nil {
+			return nil, err
+		}
+
+		return boolType, nil
+
+	case *ParenExpr:
+		return TypeCheckExpr(t.Expr)
+
+	case *ComparisonExpr:
+		return typeCheckComparisonExpr(t)
+
+	case *RangeCond:
+		return typeCheckRangeCond(t)
+
+	case *NullCheck:
+		_, err := TypeCheckExpr(t.Expr)
+		return boolType, err
+
+	case *ExistsExpr:
+		// The subquery within the exists should have been executed before
+		// expression evaluation and the exists nodes replaced with the result.
+
+	case BytesVal:
+		return stringType, nil
+
+	case StrVal:
+		return stringType, nil
+
+	case IntVal:
+		return intType, nil
+
+	case NumVal:
+		return floatType, nil
+
+	case BoolVal:
+		return boolType, nil
+
+	case ValArg:
+		// Placeholders should have been replaced before expression evaluation.
+
+	case *QualifiedName:
+		return nil, fmt.Errorf("qualified name \"%s\" not found", t)
+
+	case Tuple:
+		// TODO(tamird): handling of tuples leaves much to be desired. This
+		// is largely a consequence of go's type system not being
+		// expressive enough to encode the type of a heterogenous tuple
+		// (dBool, dString) as distinct from the tuple (dBool). We'll have
+		// to write our own type system here to do this properly.
+		return tupleType, nil
```

Comment:
I think that using `reflect.Type` as the return value from `TypeCheckExpr` is the root of the problem. I think it would be better for `TypeCheckExpr` to return a `Datum`, the type of which is the type of the expression. Then you could return the tuple itself here after checking that all of the tuple elements have compatible types (i.e. are either `dNull` or the same type).